specs/vep-1416-vdk-troubleshooting-tools/README.md

```diff
@@ -0,0 +1,101 @@
+
+To allow for easier troubleshooting of errors with data jobs, a special plugin will be introduced, which will provide capabilities to do a thread dump and either log the contents or stored locally, where it can be examined by the owner of the data job or by the team responsible for the VDK deployment. The plugin could also be extended in the future with other debugging capabilities.
+
+
+## Requirements and goals
+
+-----
+### Goals
+
+* **Introduce a vdk-jobs-troubleshooting plugin**
+
+* **Provide capability to do a thread dump**
+  - An administrator wants to get a thread dump of the data job process for troubleshooting purposes. For example, a user deploys a data job and after some time the job starts failing without any apparent reason. The administrator needs to be able to get a thread dump, so they are able to properly investigate the root cause for the issue.
+
+### Non-Goals
+
+* **Additional troubleshooting tools are not planned as part of this proposal.**
+* **The thread-dump feature would not be exposed to users/data job owners.**
+
+
+## High-level design
+
+-----
+
+![plugin_diagram.png](diagrams/plugin_diagram.png)
+
+For the proposed design, a vdk-jobs-troubleshooting plugin will be introduced. The plugin will start a http server instance, which will allow users to obtain a thread dump (which will also be printed in the logs) for further troubleshooting.
+
+
+## API design
+
+-----
+
+No changes to the public API.
+
+
+## Detailed design
+
+-----
+
+### vdk-jobs-troubleshooting Plugin
+The plugin will act as a toolbox, where data job troubleshooting tools will be implemented. As part of this proposal, only a thread-dump utility will be implemented. It will start a local web server, which will attach to port 8080 and, depending on configuration, will allow for the dump to be logged as part of normal logging processes, or a user can do port-forwarding to the data job pod and examine the contents of the thread data.
+
+![plugin_details.png](diagrams/plugin_details.png)
+
+The main configuration variable for this plugin would be:
+* VDK_TROUBLESHOOT_UTILITIES_TO_USE, which will accept a comma-separated list of string literals with the troubleshooting utilities that will be used, e.g., `"utility1,utility2,..."`.
+
+
+### Availability
+The plugin will be part of the vdk installation, so the same availability constraints apply. The http server will be running in a separate process, but it will still be available as long as the data job pod is running. In case of issues with the logging service, the http server would still be able to log the thread-dump locally in the data job pod.
```

Comment:
is the HTTP server a separate process? Why is it visualised within the VDK box, what is the VDK box?

Reply:
The VDK box refers to the VDK container in the Data Job kubernetes pod. The http server is a separate process, but it is confined within the container. I realise that the diagram is kind of misleading, in the sense that the vdk container is an init container, and the "main" container is the data job one, so will rework the diagram and provide more explanations in a subsequent PR.